freqtrade/freqtradebot.py

```diff
@@ -1140,33 +1141,78 @@ def check_handle_timedout(self) -> None:
-                            logger.warning(
-                                f'Unable to emergency sell trade {trade.pair}: {exception}')
+                    self.replace_orders(order, order_obj, trade)
+
+    def handle_timedout_orders(self, order: Dict, trade: Trade) -> None:
+        """
+        Check if any orders are timed out and cancel if necessary.
+        :param order: Order dict grabbed with exchange.fetch_order()
+        :param trade: Trade object.
+        :return: None
+        """
+        if order['side'] == trade.entry_side:
+            self.handle_cancel_enter(trade, order, constants.CANCEL_REASON['TIMEOUT'])
+        else:
+            canceled = self.handle_cancel_exit(
+                trade, order, constants.CANCEL_REASON['TIMEOUT'])
+            canceled_count = trade.get_exit_order_count()
+            max_timeouts = self.config.get('unfilledtimeout', {}).get('exit_timeout_count', 0)
+            if canceled and max_timeouts > 0 and canceled_count >= max_timeouts:
+                logger.warning(f'Emergency exiting trade {trade}, as the exit order '
+                               f'timed out {max_timeouts} times.')
+                try:
+                    self.execute_trade_exit(
+                        trade, order['price'],
+                        exit_check=ExitCheckTuple(exit_type=ExitType.EMERGENCY_EXIT))
+                except DependencyException as exception:
+                    logger.warning(
+                        f'Unable to emergency sell trade {trade.pair}: {exception}')
+
+    def replace_orders(self, order: Dict, order_obj: Optional[Order], trade: Trade) -> None:
+        """
+        Check if any orders should be replaced and do so
+        :param order: Order dict grabbed with exchange.fetch_order()
+        :param order_obj: Order object.
+        :param trade: Trade object.
+        :return: None
+        """
+        analyzed_df, _ = self.dataprovider.get_analyzed_dataframe(trade.pair,
+                                                                  self.strategy.timeframe)
+        latest_candle_open_date = analyzed_df.iloc[-1]['date'] if len(analyzed_df) > 0 else None
+        latest_candle_close_date = timeframe_to_next_date(self.strategy.timeframe,
+                                                          latest_candle_open_date)
+        # Check if new candle
+        if order_obj and latest_candle_close_date.replace(tzinfo=None) > order_obj.order_date:
+            # New candle
+            proposed_rate = self.exchange.get_rate(
+                trade.pair, side='entry', is_short=trade.is_short, refresh=True)
+            adjusted_entry_price = strategy_safe_wrapper(self.strategy.readjust_entry_price,
+                                                         default_retval=proposed_rate)(
+                pair=trade.pair, current_time=datetime.now(timezone.utc),
+                proposed_rate=proposed_rate, entry_tag=trade.enter_tag,
+                side=trade.entry_side)
```

Comment:
maybe a stupid question - but isn't the most important parameter missing (`order` this would replace)?

Sure, we can get the order through `trade.orders` - but especially if you're in a DCA scenario, you'll need to quite carefully get the "right" order - using `trade.orders[0]` will not do it then...

The doc sample even shows it as well (at least partially) - you're using `trade.open_date_utc` ... but that'll only work for the first attempt - depending on what you're trying to archive - you might replace the order every 5 minutes for 30 minutes - but you'll need to know the offset from the current order - not from the initial order.

Reply:
Honestly my initial approach was targetting just the first buy of the trade. From this perspective I've named the method to have `entry_price`. It's becoming obvious that this would be useful to be paired with DCA mode as well so I'll have to give that some though.
As for `order` as parameter I'm really not sure if the user would want that object available or not. Again maybe in the context of DCA it makes more sense to have it as well. Since the cancelation and new order are handled internally I feel it doesn't really matter for the implementation of this feature so whatever we choose to share through the strategy interface is fine by me. So I'd say that this should ultimately be your decision and I'm looking forward for you feedback on it.

Reply:
you may be targetting the first order - but there's no limitation in the code - so it'll anyway apply for all orders.

The point of sending the order with it is to have the prior price available (as well as the last adjustment date).
if you assume the same order needs 2-3 adjustments until it fills - then you'll need the time since last order - not since initial order. DCA will also benefit from this - but it's not the main concern.

Reply:
As always, thanks for the feedback!

> you may be targetting the first order - but there's no limitation in the code - so it'll anyway apply for all orders.
Indeed there was no code to apply this limitation (in live/dry modes). Calling it a fortunate mistake. Moving forward I think the desired functionality is to check/replace/cancel all open entry orders (either partially of fully unfilled).

> if you assume the same order needs 2-3 adjustments until it fills - then you'll need the time since last order - not since initial order.
This example was not a usecase I've considered, but I do see how it would add extra flexibility for the end user. Thus,
I've added the Order object to the input param list through: [3166739](https://github.com/freqtrade/freqtrade/pull/6692/commits/3166739ec9a1f211c72f922b8157619d56538148)